tests/test_distribution.py

```diff
@@ -0,0 +1,216 @@
+import os
+import json
+import unittest
+
+try:
+    from unittest import mock
+except ImportError:
+    # `Python 2` or lower than `Python 3.3` does not
+    # have the `unittest.mock` module built-in
+    import mock
+from pythonforandroid.bootstrap import Bootstrap
+from pythonforandroid.distribution import Distribution
+from pythonforandroid.recipe import Recipe
+from pythonforandroid.util import BuildInterruptingException
+from pythonforandroid.build import Context
+
+dist_info_data = {
+    "dist_name": None,
+    "bootstrap": "sdl2",
+    "archs": ["armeabi", "armeabi-v7a", "x86", "x86_64", "arm64-v8a"],
+    "ndk_api": 21,
+    "use_setup_py": False,
+    "recipes": ["hostpython3", "python3", "sdl2", "kivy", "requests"],
+    "hostpython": "/some/fake/hostpython3",
+    "python_version": "3.7",
+}
+
+
+class TestDistribution(unittest.TestCase):
+    def setUp(self):
+        self.ctx = Context()
+        self.ctx.ndk_api = 21
+        self.ctx.android_api = 27
+        self.ctx._sdk_dir = "/opt/android/android-sdk"
+        self.ctx._ndk_dir = "/opt/android/android-ndk"
+        self.ctx.setup_dirs(os.getcwd())
+        self.ctx.recipe_build_order = [
+            "hostpython3",
+            "python3",
+            "sdl2",
+            "kivy",
+        ]
+
+    def setUp_distribution_with_bootstrap(self, bs, **kwargs):
+        # extend the setUp by configuring a distribution, because some test
+        # needs a distribution to be set to be properly tested
+        self.ctx.bootstrap = bs
+        self.ctx.bootstrap.distribution = Distribution.get_distribution(
+            self.ctx,
+            name=kwargs.pop("name", "test_prj"),
+            recipes=kwargs.pop("recipes", ["python3", "kivy"]),
+            **kwargs
+        )
+
+    def tearDown(self):
+        self.ctx.bootstrap = None
+
+    def test_properties(self):
+        self.setUp_distribution_with_bootstrap(
+            Bootstrap().get_bootstrap("sdl2", self.ctx)
+        )
+        distribution = self.ctx.bootstrap.distribution
+        self.assertEqual(self.ctx, distribution.ctx)
+        expected_repr = (
+            "<Distribution: name test_prj with recipes (python3, kivy)>"
+        )
+        self.assertEqual(distribution.__str__(), expected_repr)
+        self.assertEqual(distribution.__repr__(), expected_repr)
+
+    @mock.patch("pythonforandroid.distribution.exists")
+    def test_folder_exist(self, mock_exists):
+
+        self.setUp_distribution_with_bootstrap(
+            Bootstrap().get_bootstrap("sdl2", self.ctx)
+        )
+        self.ctx.bootstrap.distribution.folder_exists()
+        mock_exists.assert_called_with(
+            self.ctx.bootstrap.distribution.dist_dir
+        )
```

Comment:
Just a general note and personal taste.
I usually also verify the number of times it was called. So in this case it would be `assert_called_once_with()`.
Also I actually prefer to use `call_args_list()` and manually assert myself because I feel like it's easier for the human eye to catch where the assertion is.
Again I'm just sharing some personal taste for the sake of sharing it, but I don't feel like one is superior over the other

Reply:
@AndreMiras, 
First of all, **so many thanks to review those tests** you are awesome :+1:, writing and reviewing those `unittest` is not an easy task because enforce us to keep the eyes in more than one file at once and all the `mock.patch` stuff it's sometimes not easy to follow. About `assert_called_once`, you are right, I should check the number of calls we make to those functions, I will address this in another pr.

**Note:** I prefer to use `assert_called_once_with` because we need less code to have the same result